### PR TITLE
Sock addr hook callout test

### PR DESCRIPTION
**Pull request by shankarseal**
## Description

Adds tests for sock_addr hook. This fixes #871 .

- Refactored common socket related classes from `xdp_tests.cpp` to a new `socket_helper.cpp` module. 
- Fix some WFP related bugs in bind and xdp hooks. The callouts were doing "soft block".
- Remove `unit_tests.exe` from selfhost runner tests as it was redundant.

## Testing

Added tests for eBPF programs handling TCP v4, v6 and UDP v4. v6 egress/ingress connections.

## Documentation

Minor documentation changes.

### CODE REVIEW

File: tests/libs/util/socket_helper.cpp [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+    : _base_socket{_sock_type, _protocol, _port}
+{}
+
+_datagram_sender_socket::_datagram_sender_socket(int _sock_type, int _protocol, uint16_t _port)
+    : _sender_socket{_sock_type, _protocol, _port}
+{
+    if (!(sock_type == SOCK_DGRAM || sock_type == SOCK_RAW) &&
+        !(protocol == IPPROTO_UDP || protocol == IPPROTO_IPV4 || protocol == IPPROTO_IPV6))
+        FAIL("datagram_socket only supports these combinations (SOCK_DGRAM, IPPROTO_UDP) and (SOCK_RAW, "
+             "IPPROTO_IPV4/IPV6)");
```
Comment (dthaler) [inline tests/libs/util/socket_helper.cpp:83]: this doesn't match what lines 80-81 do.
For example 80-81 will accept (SOCK_RAW, IPPROTO_UDP).

File: tests/libs/util/socket_helper.cpp [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+        if (timeout_expected)
+            FAIL("Receiver socket received a message when timeout was expected.");
+
+        if (!WSAGetOverlappedResult(
+                socket,
+                &overlapped,
+                reinterpret_cast<LPDWORD>(&bytes_received),
+                FALSE,
+                reinterpret_cast<LPDWORD>(&recv_flags)))
+            FAIL("WSArecvFrom on the receiver socket failed with error: " << WSAGetLastError());
```
Comment (dthaler) [inline tests/libs/util/socket_helper.cpp:192]:
```suggestion
            FAIL("WSARecvFrom on the receiver socket failed with error: " << WSAGetLastError());
```

File: tests/libs/util/socket_helper.cpp [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+            if (!timeout_expected)
+                FAIL("Receiver socket did not receive any message in 1 second.");
+        } else {
+            FAIL("Waiting on receiver socket failed with " << error);
+        }
+    }
+}
+
+void
+_receiver_socket::get_received_message(uint32_t& message_size, _Out_ char*& message)
```
Comment (dthaler) [inline tests/libs/util/socket_helper.cpp:205]: i think message should be something like `_Outptr_result_z_`

File: tests/libs/util/socket_helper.cpp [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+}
+
+_datagram_receiver_socket::_datagram_receiver_socket(int _sock_type, int _protocol, uint16_t _port)
+    : _receiver_socket{_sock_type, _protocol, _port}, sender_address{}, sender_address_size(sizeof(sender_address))
+
+{
+    if (!(sock_type == SOCK_DGRAM || sock_type == SOCK_RAW) &&
+        !(protocol == IPPROTO_UDP || protocol == IPPROTO_IPV4 || protocol == IPPROTO_IPV6))
+        FAIL("datagram_socket only supports these combinations (SOCK_DGRAM, IPPROTO_UDP) and (SOCK_RAW, "
+             "IPPROTO_IPV4/IPV6)");
```
Comment (dthaler) [inline tests/libs/util/socket_helper.cpp:218]: doesn't match what 215-216 does, which also accept (SOCK_RAW, IPPROTO_UDP). and https://docs.microsoft.com/en-us/windows/win32/winsock/tcp-ip-raw-sockets-2 explicitly says windows supports that combo

File: tests/libs/util/socket_helper.cpp [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+
+    if (error != 0)
+        FAIL("Obtaining AcceptEx function pointer failed with " << WSAGetLastError());
+
+    // Post listen.
+    listen(socket, SOMAXCONN);
+
+    // Create accept socket.
+    accept_socket = WSASocket(AF_INET6, sock_type, protocol, nullptr, 0, WSA_FLAG_OVERLAPPED);
+    uint32_t ipv6_opt = 0;
```
Comment (dthaler) [inline tests/libs/util/socket_helper.cpp:286]:
```suggestion
    uint32_t ipv6_option = 0;
```

File: tests/libs/util/socket_helper.cpp [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+            reinterpret_cast<LPDWORD>(&bytes_received),
+            &overlapped)) {
+        int wsaerr = WSAGetLastError();
+        if (wsaerr != WSA_IO_PENDING)
+            FAIL("AcceptEx failed with " << wsaerr);
+    }
+}
+
+void
+_stream_receiver_socket::get_sender_address(PSOCKADDR& from, int& from_length)
```
Comment (dthaler) [inline tests/libs/util/socket_helper.cpp:322]:
```suggestion
_stream_receiver_socket::get_sender_address(_Out* PSOCKADDR& from, _Out_ int& from_length)
```

File: tests/libs/util/socket_helper.cpp [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+{
+    from = (PSOCKADDR)(recv_buffer.data() + message_length);
```
Comment (dthaler) [inline tests/libs/util/socket_helper.cpp:324]: this looks like a bug to me since a SOCKADDR_IN6 is larger than SOCKADDR so this will not copy from_length bytes.

File: scripts/run_tests.psm1 [hunk 8]
```diff
@@ -76,7 +76,6 @@ function Invoke-CICDTests
     try {
 
          $TestList = @(
-            "unit_tests.exe",
```
Comment (dthaler) [inline scripts/run_tests.psm1:79]: why this change?

File: docs/GettingStarted.md [hunk 9]
```diff
@@ -75,6 +75,7 @@ This will build the following binaries:
 * `sample_ext_app.exe`: A sample application for testing the sample extension driver.
 * `xdp_tests.exe`: Application for testing various XDP functionalities.  This requires the EbpSvc service to be running,
                 and the EbpCore and NetEbpfExt drivers to be loaded on a remote system to test.
+* `socket_tests.exe`: Application for testing hook for BPF_CGROUP_SOCK_ADDR program type and related attach types.
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply docs/GettingStarted.md:78]: I dont know why I keep missing the "the" !

File: netebpfext/net_ebpf_ext_sock_addr.c [hunk 10]
```diff
@@ -358,8 +358,8 @@ _net_ebpf_extension_sock_addr_copy_wfp_connection_fields(
             incoming_values[destination_ip_address_field].value.byteArray16,
             sizeof(FWP_BYTE_ARRAY16));
     }
-    sock_addr_ctx->msg_src_port = incoming_values[source_port_field].value.uint16;
-    sock_addr_ctx->user_port = incoming_values[destination_port_field].value.uint16;
+    sock_addr_ctx->msg_src_port = _byteswap_ushort(incoming_values[source_port_field].value.uint16);
+    sock_addr_ctx->user_port = _byteswap_ushort(incoming_values[destination_port_field].value.uint16);
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply netebpfext/net_ebpf_ext_sock_addr.c:362]: Yes WFP gives these in host byte order and has to be swapped. I will see if I can use `htonl` instead.

File: scripts/run_tests.psm1 (hunk 8 above)
Comment (shankarseal) [reply scripts/run_tests.psm1:79]: unit_tests run in other WFs. Its redundant in the driver VM runs. Besides these tests dont invoke the driver code.

File: tests/libs/util/socket_helper.cpp (hunk 7 above)
Comment (shankarseal) [reply tests/libs/util/socket_helper.cpp:324]: There is no copy here. The function returns a pointer to a buffer that stores the remote address.

File: netebpfext/net_ebpf_ext_hook_provider.h [hunk 11]
```diff
@@ -63,6 +63,7 @@ net_ebpf_extension_hook_client_set_provider_data(_In_ net_ebpf_extension_hook_cl
  * @brief Get the hook-specific provider data for the attached client.
  *
  * @param[in] hook_client Pointer to attached hook NPI client.
+ * @param[in] data hook-specific provider data.
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply netebpfext/net_ebpf_ext_hook_provider.h]: Was a bad merge.

File: netebpfext/net_ebpf_ext.c [hunk 12]
```diff
@@ -21,6 +21,8 @@
 #include "net_ebpf_ext_bind.h"
 #include "net_ebpf_ext_sock_addr.h"
 #include "net_ebpf_ext_xdp.h"
+#include "net_ebpf_ext_sock_addr.h"
+#include "net_ebpf_ext_xdp.h"
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply netebpfext/net_ebpf_ext.c]: Was a bad merge.

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (shankarseal) [reply netebpfext/net_ebpf_ext_sock_addr.c]: Could not find a header that can be included in kernel driver with definition of htonl etc. So ended up #defining one. At least makes the code more readable.

File: tests/libs/util/socket_helper.cpp (hunk 1 above)
Comment (shankarseal) [reply tests/libs/util/socket_helper.cpp]: Changed the log statement.

File: scripts/run_tests.psm1 (hunk 8 above)
Comment (dthaler) [reply scripts/run_tests.psm1:79]: But this script is called "run_tests" not "run_vm_tests", and the current name implies it will run all tests.  If you don't mean to run all tests, then I recommend renaming this file.

File: docs/GettingStarted.md [hunk 13]
```diff
@@ -208,6 +209,9 @@ This application tests various XDP functionalities. It has the following tests:
    3. From a remote host, run xdp_tests.exe and in `--remote-ip` parameter pass an IPv4 or IPv6 address of an Ethernet-like interface on the system under test in string format.
    4. Unload the program from system under test by running `delete program <id>` on the netsh prompt, where <id> is the ID noted above.
 
+### socket_tests.exe
+This application loads the `cgroup_sock_addr.o` eBPF program and attaches to hooks to handle various socket operations. Currently it tests authorizing ingress and egress connections based on entry in a map passed to the program.
```
Comment (dthaler) [inline docs/GettingStarted.md:213]: grammar:
```suggestion
This application loads the `cgroup_sock_addr.o` eBPF program and attaches to hooks to handle various socket operations. Currently it tests authorizing ingress and egress connections based on a map passed to the program.
```
(either that or "based on entries in a map" with plural entries)

File: netebpfext/net_ebpf_ext.h [hunk 14]
```diff
@@ -49,6 +49,9 @@
 #define NET_EBPF_EXTENSION_POOL_TAG 'Nfbe'
 #define NET_EBPF_EXTENSION_NPI_PROVIDER_VERSION 0
 
+#define htonl(x) _byteswap_ulong(x)
+#define htons(x) _byteswap_ushort(x)
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext.h:53]: https://github.com/microsoft/ebpf-for-windows/pull/916/files#diff-6e4bda166b4c5a3103506c3378f2cfa27ee4cc3edfba5ef7dc01ec7ecfc943f1 adds bpf_endian.h, maybe we can combine this into there?

File: tests/libs/util/socket_helper.cpp (hunk 6 above)
Comment (dthaler) [reply tests/libs/util/socket_helper.cpp:322]: This one was marked as resolved without reply or change, so un-resolving... does `_Out_` not work with `&` args?  You did change socket_helper.h but not here.

File: tests/libs/util/socket_helper.cpp [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+        nullptr);
+    if (error != 0) {
+        int wsaerr = WSAGetLastError();
+        if (wsaerr != WSA_IO_PENDING)
+            FAIL("WSARecvFrom failed with " << wsaerr);
+    }
+}
+
+void
+_datagram_receiver_socket::get_sender_address(PSOCKADDR& from, int& from_length)
```
Comment (dthaler) [inline tests/libs/util/socket_helper.cpp:252]:
```suggestion
_datagram_receiver_socket::get_sender_address(_Out_ PSOCKADDR& from, _Out_ int& from_length)
```

File: tests/libs/util/socket_helper.h [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+ * @class A dual stack UDP or raw socket bound to wildcard address that is used to receive datagrams.
+ */
+typedef class _base_socket
+{
+  public:
+    _base_socket(int _sock_type, int _protocol, uint16_t port);
+    ~_base_socket();
+
+    void
+    get_local_address(PSOCKADDR& address, int& address_length);
```
Comment (dthaler) [inline tests/libs/util/socket_helper.h:36]:
```suggestion
    get_local_address(_Out_ PSOCKADDR& address, _Out_ int& address_length);
```

File: tests/libs/util/socket_helper.cpp [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+            if (!timeout_expected)
+                FAIL("Receiver socket did not receive any message in 1 second.");
+        } else {
+            FAIL("Waiting on receiver socket failed with " << error);
+        }
+    }
+}
+
+void
+_receiver_socket::get_received_message(uint32_t& message_size, _Outptr_result_z_ char*& message)
```
Comment (dthaler) [inline tests/libs/util/socket_helper.cpp:205]:
```suggestion
_receiver_socket::get_received_message(_Out_ uint32_t& message_size, _Outptr_result_z_ char*& message)
```

File: tests/libs/util/socket_helper.cpp (hunk 6 above)
Comment (shankarseal) [reply tests/libs/util/socket_helper.cpp]: I only updated the .h file and didnt update the .cpp. Of course static analysis failed due to this. Fixed now.

File: scripts/run_tests.psm1 (hunk 8 above)
Comment (shankarseal) [reply scripts/run_tests.psm1:79]: renamed file.

File: netebpfext/net_ebpf_ext.h (hunk 14 above)
Comment (shankarseal) [reply netebpfext/net_ebpf_ext.h:53]: I have added bpf_htonl etc. in the new `bpf_endian.h` file. Those maybe used by bpf programs. But as far as the netebpfext.sys driver is concerned, since it targets windows only, IMO its OK to use DDK functions such as rtlswapbytes etc.

File: docs/GettingStarted.md [hunk 18]
```diff
@@ -75,6 +75,7 @@ This will build the following binaries:
 * `sample_ext_app.exe`: A sample application for testing the sample extension driver.
 * `xdp_tests.exe`: Application for testing various XDP functionalities.  This requires the EbpSvc service to be running,
                 and the EbpCore and NetEbpfExt drivers to be loaded on a remote system to test.
+* `socket_tests.exe`: Application for testing the hook for the BPF_CGROUP_SOCK_ADDR program type and related attach types.
```
Comment (dthaler) [inline docs/GettingStarted.md:78]:
```suggestion
* `socket_tests.exe`: Application for testing the BPF_CGROUP_SOCK_ADDR program type and related attach types.
```
("hook" should be plural, but simplified sentence instead of just making it be plural)

**Review by dthaler — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]
